db/commitlog/commitlog.cc

```diff
@@ -904,41 +918,47 @@ class db::commitlog::segment : public enable_shared_from_this<segment>, public c
         }
 
         if (_closed) {
-            return make_exception_future<rp_handle>(std::runtime_error("commitlog: Cannot add data to a closed segment"));
+            return make_exception_future<>(std::runtime_error("commitlog: Cannot add data to a closed segment"));
         }
 
         buf_memory -= permit.release();
         _segment_manager->account_memory_usage(buf_memory);
 
-        replay_position rp(_desc.id, position());
-        _cf_dirty[id]++; // increase use count for cf.
+        for (size_t entry = 0; entry < writer->num_entries; ++entry) {
+            replay_position rp(_desc.id, position());
+            auto id = writer->id(entry);
+            auto entry_size = writer->num_entries == 1 ? size : writer->size(*this, entry);
+            auto es = entry_size + entry_overhead_size;
 
-        rp_handle h(static_pointer_cast<cf_holder>(shared_from_this()), std::move(id), rp);
+            _cf_dirty[id]++; // increase use count for cf.
 
-        auto out = _buffer_ostream.write_substream(s);
-        crc32_nbo crc;
+            rp_handle h(static_pointer_cast<cf_holder>(shared_from_this()), std::move(id), rp);
 
-        write<uint32_t>(out, s);
-        crc.process(uint32_t(s));
-        write<uint32_t>(out, crc.checksum());
+            auto out = _buffer_ostream.write_substream(es);
+            crc32_nbo crc;
 
-        // actual data
-        auto entry_out = out.write_substream(size);
-        auto entry_data = entry_out.to_input_stream();
-        writer->write(*this, entry_out);
-        entry_data.with_stream([&] (auto data_str) {
-            crc.process_fragmented(ser::buffer_view<typename std::vector<temporary_buffer<char>>::iterator>(data_str));
-        });
+            write<uint32_t>(out, es);
+            crc.process(uint32_t(es));
+            write<uint32_t>(out, crc.checksum());
+
+            // actual data
+            auto entry_out = out.write_substream(entry_size);
+            auto entry_data = entry_out.to_input_stream();
+            writer->write(*this, entry_out, entry);
+            entry_data.with_stream([&] (auto data_str) {
+                crc.process_fragmented(ser::buffer_view<typename std::vector<temporary_buffer<char>>::iterator>(data_str));
+            });
+
+            write<uint32_t>(out, crc.checksum());
```

Comment:
Looks like we checksum each mutation separately. How can we guaranty that the write is atomic?

Reply:
The CRC is still inside the segment buffer, and in the uninterruptible write loop. The guarantee is that all the data is committed (i.e. flushed to disk) together. Since we reserve the segment space and fill all the entries inside this buffer in a single call, we either do them all, or none. 
_Reading_ will not be atomic, in the sense that we will read each entry separately, but @tgrabiec  said this was fine.

Reply:
I am not sure about the guaranty. If a buffer is large and the crash happens while it is flushed to a disk why can't it be written partially? We have crc exactly to check for such cases, no?

Reply:
Well, any disk corruption, but yes. You can only really trust a multi-write if you use it in direct flush (or batch) mode, so you know things are actually on disk. 
However, as stated above, pushing all the entries into a single segment entry (crc block) does not work well either, since then you would have a single RP (handle) representing this, and can't properly signal from respective table when we no longer need this segment. Can be dealt with, but I am doubtful it would look nicer...

Reply:
The system may crash before flush completes. @tgrabiec what is the guaranty you want from that? Because as it is implemented I do not see a difference between this and writing mutations one by one.

Reply:
The only difference is that you are guaranteed to be written in a single flush call, not sequential ones. 
If you want additional crc-guarantees I would probably have to evolve the disk format to allow for nested entries or something.

Reply:
We need atomicity in case of a crash, so I think the CRC problem mentioned by Gleb should be fixed.

Reply:
Very well. This requires file format change, but that is what version numbers are for. Now we explicitly wrap the multi-entry in a header + crc and verify it on read.